src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -0,0 +1,476 @@
+        }
+
+        // NOTE!!
+        // This is a copy of C++ implementation in CastCache.cpp
+        // Keep the copies, if possible, in sync.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static CastResult TryGet(nuint source, nuint target)
+        {
+            const int BUCKET_SIZE = 8;
+            int[]? table = s_table;
+
+            // we use NULL as a sentinel for a rare case when a table could not be allocated
+            // because we avoid OOMs.
+            // we could use 0-element table instead, but then we would have to check the size here.
+            if (table != null)
+            {
+                int index = KeyToBucket(table, source, target);
+                for (int i = 0; i < BUCKET_SIZE;)
+                {
+                    ref CastCacheEntry pEntry = ref Element(table, index);
+
+                    // must read in this order: version -> entry parts -> version
+                    // if version is odd or changes, the entry is inconsistent and thus ignored
+                    int version = Volatile.Read(ref pEntry._version);
+                    nuint entrySource = pEntry._source;
+
+                    // mask the lower version bit to make it even.
+                    // This way we can check if version is odd or changing in just one compare.
+                    version &= ~1;
+
+                    if (entrySource == source)
+                    {
+                        nuint entryTargetAndResult = Volatile.Read(ref pEntry._targetAndResult);
+                        // target never has its lower bit set.
+                        // a matching entryTargetAndResult would the have same bits, except for the lowest one, which is the result.
+                        entryTargetAndResult ^= target;
+                        if (entryTargetAndResult <= 1)
+                        {
+                            if (version != pEntry._version)
+                            {
+                                // oh, so close, the entry is in inconsistent state.
+                                // it is either changing or has changed while we were reading.
+                                // treat it as a miss.
+                                break;
+                            }
+
+                            return (CastResult)entryTargetAndResult;
+                        }
+                    }
+
+                    if (version == 0)
+                    {
+                        // the rest of the bucket is unclaimed, no point to search further
+                        break;
+                    }
+
+                    // quadratic reprobe
+                    i++;
+                    index = (index + i) & TableMask(table);
+                }
+            }
+            return CastResult.MaybeCast;
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object JITutil_IsInstanceOfAny_NoCacheLookup(void* toTypeHnd, object obj);
```

Comment:
Nit: I would use regular managed naming convention for these. `JITutil_` prefix made sense when it was C entrypoint  or written in assembly. It does not make sense not that it is a regular method embedded within a type.

Reply:
I figured it would be helpful in reviews - to see where the code moved to. But I can drop the "JITUtil_" part. 

---
In reply to: [368261375](https://github.com/dotnet/runtime/pull/1068#discussion_r368261375) [](ancestors = 368261375)

Reply:
Should we remove the `JIT_` in the entry points?  Those _are_ supposed to be called by JIT only.

---
In reply to: [368261476](https://github.com/dotnet/runtime/pull/1068#discussion_r368261476) [](ancestors = 368261476,368261375)

Reply:
I would remove `JIT_`. The enclosing type name `CastHelpers` is descriptive enough.

`JIT_` prefix actually misleads people into thinking that this method is part of the JIT. It was particularly confusing for `JIT_Throw`. The JIT team had to explain all the time that `JIT_Throw` on the stack does not imply that the crash is caused by the bug in the JIT. We have renamed it to `IL_Throw` to eliminate the confusion.